### PR TITLE
#1319 Add table module pattern

### PR DESCRIPTION
**Pull request by tao-sun2**
- Resolves issue #1319 
- In Table Module a single class encapsulates all the domain logic for all records stored in a table or view.
- I use the same scenario which is used in [Transaction Script](https://github.com/iluwatar/java-design-patterns/tree/master/transaction-script) and refactor the code from Transaction Script pattern to Table Module pattern.

### CODE REVIEW

**Comment by tao-sun2** [comment]
The duplication rate is high because I use the same scenario as Transaction Script and refactor the code from one pattern to another. Should I choose a new scenario and reimplement the example?

**Comment by ohbus** [comment]
> The duplication rate is high because I use the same scenario as Transaction Script and refactor the code from one pattern to another. Should I choose a new scenario and reimplement the example?

Yes please

**Review by ohbus — CHANGES_REQUESTED** [review]
While you are already working on fixing the code duplications please ensure that other parameters like **code smells** and **bugs** are also removed in the process.

**Comment by tao-sun2** [comment]
I will try to do this.

File: table-module/src/main/java/com/iluwatar/tablemodule/RoomDaoImpl.java [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+   */
+  @Override
+  public Boolean delete(final Room room) throws Exception {
+    try (var connection = getConnection();
+         var statement = connection.prepareStatement("DELETE FROM ROOMS WHERE ID = ?")) {
+      statement.setInt(1, room.getId());
+      return statement.executeUpdate() > 0;
+    } catch (Exception e) {
+      throw new Exception(e.getMessage(), e);
+    }
```
Comment (iluwatar) [inline table-module/src/main/java/com/iluwatar/tablemodule/RoomDaoImpl.java:187]: Maybe just let it throw

Comment (iluwatar) [reply table-module/src/main/java/com/iluwatar/tablemodule/RoomDaoImpl.java:187]: Please check the same issue elsewhere too

File: table-module/src/main/java/com/iluwatar/tablemodule/RoomTableModule.java [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+   *
+   * @param roomNumber room to book
+   * @throws Exception if any error
+   */
+  public void bookRoom(final int roomNumber) throws Exception {
+
+    var room = roomDao.getById(roomNumber);
+
+    if (room.isEmpty()) {
+      throw new Exception("Room number: " + roomNumber + " does not exist");
```
Comment (iluwatar) [inline table-module/src/main/java/com/iluwatar/tablemodule/RoomTableModule.java:58]: It's bad practice to throw generic exceptions. I think we should create our own room booking exception class.

File: table-module/src/main/java/com/iluwatar/tablemodule/RoomTableModule.java [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+    } else {
+      if (room.get().isBooked()) {
+        var updateRoomBooking = room.get();
+        updateRoomBooking.setBooked(false);
+        int refundAmount = updateRoomBooking.getPrice();
+        roomDao.update(updateRoomBooking);
+
+        LOGGER.info("Booking cancelled for room number: " + roomNumber);
+        LOGGER.info(refundAmount + " is refunded");
```
Comment (iluwatar) [inline table-module/src/main/java/com/iluwatar/tablemodule/RoomTableModule.java:90]: Combine into single logging line

File: table-module/src/main/java/com/iluwatar/tablemodule/RoomDao.java [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package com.iluwatar.tablemodule;
+
+import java.util.Optional;
+import java.util.stream.Stream;
+
+public interface RoomDao {
```
Comment (iluwatar) [inline table-module/src/main/java/com/iluwatar/tablemodule/RoomDao.java:29]: Could we use some other name for the classes not to confuse with the DAO pattern?

**Review by iluwatar — CHANGES_REQUESTED** [review]

**Comment by tao-sun2** [comment]
I am deeply sorry that I closed this pull request and create a new pull request for the same issue #1319  because I reimplement the Table Module pattern instead of refactoring the code from Transaction Script pattern to Table Module pattern. The link for the new pull request is https://github.com/iluwatar/java-design-patterns/pull/1742

**Comment by iluwatar** [comment]
@tao-sun2 although you closed this pull request, please consider the review comments in the new PR. I would not want to do the same work again.

**Pull request closed without merging**